File: static/CWE-22/java/patched.java
// Remediation for CWE-22: Path Traversal
// Fix: Canonicalize the resolved path and verify it stays within the trusted base directory.

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

class Patched_22 {
    private static final Path BASE = Paths.get("/var/www/uploads").toAbsolutePath().normalize();

    public byte[] readFile(String userInput) throws Exception {
        File f = new File(BASE.toFile(), userInput);
        Path resolved = f.getCanonicalFile().toPath();
        if (!resolved.startsWith(BASE)) {
            throw new SecurityException("Path traversal attempt detected");
        }
        return Files.readAllBytes(resolved);
    }
}

File: static/CWE-22/csharp/patched.cs
// Remediation for CWE-22: Improper Limitation of a Pathname to a Restricted Directory ('Path Traversal')
// Fix: Resolve the combined path and verify it stays within the base directory before reading.

using System;
using System.IO;

namespace GuineaPig.Sast
{
    class Patched_22
    {
        public string ReadFile(string userInput)
        {
            string baseDir = Path.GetFullPath("/var/www/files");
            string fullPath = Path.GetFullPath(Path.Combine(baseDir, userInput));

            string baseWithSep = baseDir.EndsWith(Path.DirectorySeparatorChar.ToString())
                ? baseDir
                : baseDir + Path.DirectorySeparatorChar;

            if (!fullPath.StartsWith(baseWithSep, StringComparison.Ordinal))
            {
                throw new UnauthorizedAccessException("Path is outside the allowed directory.");
            }

            return File.ReadAllText(fullPath);
        }
    }
}
